File: 1461. Check If a String Contains All Binary Codes of Size K/Solution.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public boolean hasAllCodes(String s, int k) {
        if (s.length() - k + 1 < (1 << k)) {
            return false;
        }
        int n = Integer.parseInt(s.substring(0, k), 2);
        Set<Integer> seen = new HashSet<>();
        seen.add(n);
        int head_mask = ~(1 << (k - 1));
        for (int i = k; i < s.length(); ++i) {
            n = (n & head_mask) << 1 | (s.charAt(i) - '0');
            seen.add(n);
        }
        return seen.size() == (1 << k);
    }
}

File: 1461. Check If a String Contains All Binary Codes of Size K/Solution.py
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        if len(s) - k + 1 < (1 << k):
            return False
        n = int(s[:k], 2)
        seen = set([n])
        head_mask = ~(1 << (k - 1))
        for i in range(k, len(s)):
            n = (n & head_mask) << 1 | (ord(s[i]) - ord('0'))
            seen.add(n)
        return len(seen) == (1 << k)
